kratos/mpi/utilities/mpi_search_utilities.h

```diff
@@ -0,0 +1,284 @@
+     * @param AllPointsAreTheSame If all points are the same (-1 to be computed, 0 if not, 1 if yes)
+     * @param NumberOfPoints The number of points in the range
+     * @param TotalNumberOfPoints The total number of points in all the ranges
+     * @return The number of points in the range
+     * @tparam TPointIteratorType The type of the point iterator
+     */
+    template<typename TPointIteratorType>
+    static int MPISynchronousPointSynchronization(
+        TPointIteratorType itPointBegin,
+        TPointIteratorType itPointEnd,
+        std::vector<double>& rAllPointsCoordinates,
+        const DataCommunicator& rDataCommunicator,
+        int AllPointsAreTheSame = -1,
+        int NumberOfPoints = -1,
+        int TotalNumberOfPoints = -1
+        )
+    {
+        // First check that the points are the same in all processes
+        if (AllPointsAreTheSame < 0) {
+            AllPointsAreTheSame = static_cast<int>(CheckAllPointsAreTheSame(itPointBegin, itPointEnd, NumberOfPoints, TotalNumberOfPoints, rDataCommunicator));
+        }
+
+        KRATOS_DEBUG_ERROR_IF(NumberOfPoints < 0) << "The number of points is negative" << std::endl;
+        KRATOS_DEBUG_ERROR_IF(TotalNumberOfPoints < 0) << "The total number of points is negative" << std::endl;
+
+        // If all points are the same
+        if (AllPointsAreTheSame) {
+            rAllPointsCoordinates.resize(NumberOfPoints * 3);
+            std::size_t counter = 0;
+            array_1d<double, 3> coordinates;
+            unsigned int i_coord;
+            for (auto it_point = itPointBegin ; it_point != itPointEnd ; it_point++) {
+                noalias(coordinates) = it_point->Coordinates();
+                for (i_coord = 0; i_coord < 3; ++i_coord) {
+                    rAllPointsCoordinates[3 * counter + i_coord] = coordinates[i_coord];
+                }
+                ++counter;
+            }
+        } else { // If not
+            // MPI information
+            const int world_size = rDataCommunicator.Size();
+
+            // Getting global number of points
+            std::vector<int> points_per_partition(world_size);
+            std::vector<int> send_points_per_partition(1, NumberOfPoints);
+            rDataCommunicator.AllGather(send_points_per_partition, points_per_partition);
+
+            // Getting global coordinates
+            rAllPointsCoordinates.resize(TotalNumberOfPoints * 3);
+            std::vector<double> send_points_coordinates(NumberOfPoints * 3);
+            std::size_t counter = 0;
+            array_1d<double, 3> coordinates;
+            unsigned int i_coord;
+            for (auto it_point = itPointBegin ; it_point != itPointEnd ; it_point++) {
+                noalias(coordinates) = it_point->Coordinates();
+                for (i_coord = 0; i_coord < 3; ++i_coord) {
+                    send_points_coordinates[3 * counter + i_coord] = coordinates[i_coord];
+                }
+                ++counter;
+            }
```

Comment:
Note that, if you move the calculation of the number of points outside the check function all this could be outside the if, and the only branching code will be the creation of the buffers and the `sendrecv`:

```c++

...
        send_points_coordinates[3 * counter + i_coord] = coordinates[i_coord];
....
if (AllPointsAreTheSame) {
    rAllPointsCoordinates = send_points_coordinates
} else {
   // What you have below...
}
```

Reply:
Okay, true